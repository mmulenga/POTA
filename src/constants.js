/**
 * This file contains all string constants for the project
 * Use this instead of hardcoding string values
 * HTML and one-time-use messages are an acceptable exception
 */

/* Conditions */
export const Condition = {
  /* Cardiovascular Diseases */
  AtrialFib: 'Atrial fibrillation / History of irregular heart beat',
  Defib: 'Defibrillator / Pacemaker',
  CAD: 'Coronary artery disease / Cardiac stent',
  CerebralDisease: 'History of cerebral vascular disease / TIA / Stroke',
  VHD: 'Valvular heart disease / Valve Replacement',
  HeartFail: 'Heart failure',
  PVD: 'Peripheral Vascular disease (PVD)',
  /* Pulmonary Diseases */
  PulmonaryVascular: 'History of pulmonary vascular disease',
  PulmDisease: 'Severe COPD, Home Oxygen, Pulmonary HTN',
  /* Other Diseases */
  Risk: 'Age ≥ 69 years old OR ≥ 2 risk factors (HTN, DM, OSA, CKD, BMI > 35)',
  Bleeding: 'Bleeding disorders',
  Anemia: 'History of anemia in past 12 months or ongoing blood loss',
  KidneyDisease: 'Chronic kidney disease / on dialysis',
  Diabetes: 'Diabetes',
  Malignancy: 'Present malignancy / surgery for malignancy',
  Hepatic: 'Hepatic disease',
  Adrenal: 'History of adrenal disease',
  Pituitary: 'History of pituitary disease',
  Endocrine: 'History of major systemic endocrine disease',
  Thyroid: 'Known thyroid disease or on thyroid supplementation',
  Digoxin: 'Digoxin',
  Lithium: 'Lithium',
  Diuretics: 'Diuretics',
  ACEI: 'ACE-I',
  ARB: 'ARB',
  NSAIDS: 'NSAIDS',
  Anticoagulant: 'Anticoagulants',
  Antiplatelet: 'Antiplatelet (ASA excluded)',
  Steroid: 'Systemic steroid use within 6 months',
};

/* Medical Examinations */
export const Exam = {
  GnS: 'Group and Screen',
  ECG: 'ECG',
  CBC: 'CBC',
  RenPanel: 'Renal Panel (Creat + Lytes)',
  PTTINR: 'PTT/INR',
  LFT: 'LFT',
  Gluc: 'Gluc',
  HbA1C: 'HbA1C',
  CXR: 'CXR',
  TSH: 'TSH',
};

/* Glossary Entries */
export const Glossary = {
  AtrialFib: 'No extra information.',
  Defib: 'No extra information.',
  CAD: 'No extra information.',
  CardiacStent: 'No extra information.',
  CerebralDisease: 'No extra information.',
  PulmonaryVascular: 'No extra information.',
  VHD: 'No extra information.',
  HeartFail: 'No extra information.',
  PVD: 'No extra information.',
  /* Pulmonary Diseases */
  PulmDisease: '• Spirometric values AND symptomatic assessment should be sought. \n' +
               '• Post Bronchodilator FEV1: 30 - 50% of predicted; FEV1/FVC < 0.70 \n' +
               '• History of exacerbations (including prior hospitalizations) should be recorded.\n' +
               '• Worsening air flow limitation, greater SOB, reduced exercise capacity, having an ' +
               'impact on quality of life, and ability to complete activities of daily living should be present. \n',
  /* Other Diseases */
  Risk: 'No extra information.',
  Bleeding: '• Hemophiliac. \n' +
            '• History of extensive investigation due to a bleeding episode. \n' +
            '• History of extensive bleeding needing transfusion in previous surgeries. \n' +
            '• Bruising easily more than usual. \n',
  Anemia: 'No extra information.',
  ActiveBleeding: 'No extra information.',
  KidneyDisease: 'No extra information.',
  Renal: '(AKI). \n' +
         '• Chronic kidney disease / on dialysis. \n' +
         '• Diabetes. \n' +
         '• Heart failure. \n' +
         '• Age ≥ 65. \n' +
         '• Liver disease. \n' +
         '• Use of drugs with nephrotoxic potential in the perioperative period (in particular NSAIDs after surgery). \n',
  Diabetes: 'No extra information.',
  Malignancy: 'No extra information.',
  Hepatic: '• Alcohol abuse (i.e., female: 14 drinks / week; male: 21 drinks / week). \n' +
           '• Jaundice due to liver disease. \n' +
           '• All forms of hepatitis. \n' +
           '• Cirrhosis; Hepatic cancer. \n',
  Adrenal: 'No extra information.',
  Pituitary: '• Acromegaly. \n' +
             '• Cushing’s disease. \n' +
             '• Sheehan syndrome. \n' +
             '• Growth hormone deficiency. \n' +
             '• Hyper/hypo pituitarism. \n',
  Endocrine: '(Excluding Diabetes) \n' +
             '• Parathyroid disease \n' +
             '• Addison’s disease \n' +
             '• Cushing’s syndrome \n' +
             '• Graves’ disease \n',
  Thyroid: 'No extra information.',
  /* Medications */
  Digoxin: 'No extra information.',
  Lithium: 'No extra information.',
  Diuretics: 'Furosemide; Hydrochlorothiazide; Acetazolamide; Spironolactone; Chlorthalidone',
  ACEI: 'Captopril; Enalapril; Ramipril; Trandolapril; Perindopril',
  ARB: 'Valsartan; Candesartan; Irbesartan; Telmisartan; Losartan; Olmesartan',
  NSAIDS: 'No extra information.',
  Anticoagulant: 'Warfarin; Heparins; Dabigatran; Apixaban; Rivaroxaban',
  Antiplatelet: 'Clopidogrel; Ticagrelor; Prasugrel; Aspirin/Dipyridamole (Aggrenox)',
  Steroid: 'Prednisone; Prednisolone; Methylprednisolone; Hydrocortisone; Dexamethasone; Oral Budesonide',
};
