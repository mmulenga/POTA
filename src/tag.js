import { assert } from '@/main';
import { Condition, Exam } from '@/constants';

/**
 * This is the internal collection where all exam information will be saved to.
 * Do not export this or attempt to access it externally!
 */
const examCollection = {};
examCollection.exams = {};
examCollection.conditionalExams = {};
examCollection.validityPeriods = {};

/**
 * Adds a new global exam to the system.
 * Not to be called outside this file!
 * @param {String} name The name of the condition
 * @param {String[]} exams The name of the exams required
 * Preconditons: Name and Exams must all be from the Constants file
 * Postconditions: The global examCollection object will be extended to include the new exams
 */
function addExams(name, exams) {
  assert(Object.values(Condition).indexOf(name) >= 0, `Unknown condition ${name} used for Tag`);
  for (let i = 0; i < exams.length; i += 1) {
    assert(Object.values(Exam).indexOf(exams[i]) >= 0, `Unknown exam ${exams[i]} used for Tag`);
  }
  examCollection.exams[name] = exams;
}

/**
 * ConditionalExam is a simple structure containing a condition phrase and the
 * exams which are required if that phrase is true.
 */
export class ConditionalExam {
  constructor(conditionPhrase, exams) {
    this.conditionPhrase = conditionPhrase;
    this.exams = exams;
  }
}

/**
 * Adds a new global exam to the system.
 * Not to be called outside this file!
 * @param {String} name The name of the condition
 * @param {Object} conditionalExams A mapping of requirements to their conditional exams
 *    Format for conditionalExams:
 *    { requirementName: [examNames] }
 * Preconditons: Name and requirements/exams must all be from the Constants file
 * Postconditions: The global examCollection object will be extended to include
 *                 the new conditional exams as ConditionalExam objects
 */
function addConditionalExams(name, conditionalExams) {
  const conditionalExamList = [];
  // Iterating over keys and values
  const iter = Object.entries(conditionalExams);
  for (let i = 0; i < iter.length; i += 1) {
    const conditionPhrase = iter[i][0];
    const exams = iter[i][1];
    for (let j = 0; j < exams.length; j += 1) {
      assert(Object.values(Exam).indexOf(exams[j]) >= 0, `Unknown Exam ${exams[j]} used for tag`);
    }
    conditionalExamList.push(new ConditionalExam(conditionPhrase, exams));
  }
  examCollection.conditionalExams[name] = conditionalExamList;
}

/**
 * Adds a new global tag to the system.
 * This is to be used for setting up the knowledge base required for a Preop recommendation.
 * If the need for an examination is dependent on a certain condition, it may be passed in:
 * @param {String} name The name of the condition
 * @param {String[]} exams The names of the exams required
 * @param {Object} conditionalExams (optional) A mapping of requirements to their conditional exams
 *    Format for conditionalExams:
 *    { requirementName: [examNames] }
 * Preconditons: Name and Exams must all be from the Constants file
 * Postconditions: The global examCollection object will be extended to include the new Tag
 */
export function Tag(name, exams, conditionalExams = {}) {
  addExams(name, exams);
  addConditionalExams(name, conditionalExams);
}

/**
 * Returns a summary of all exams required for a given condition name.
 * @param {String} name
 */
export function GetExams(name) {
  assert(name in examCollection.exams, `GetExams called with unknown condition ${name}`);
  if (name in examCollection.exams) {
    return examCollection.exams[name];
  }
  return [];
}

/**
 * Returns a comprehensive summary of the entire knowledge base of the system.
 * This should be used for external access instead of directly referencing examCollection
 */
export function GetAllExams() {
  return examCollection;
}

/**
 * Returns a list of ConditionalExam objects associated with the given condition name
 */
export function GetConditionalExams(name) {
  assert(name in examCollection.conditionalExams, `GetConditionalExams called with unknown condition ${name}`);
  if (name in examCollection.conditionalExams) {
    return examCollection.conditionalExams[name];
  }
  return [];
}


/**
 * Sets the validity period for a given exam
 * Validity period is the time window before surgery in which an exam must be performed
 * @param {String} exam The name of the exam
 * @param {String} validityString A string of the format "x months" or "x days"
 * Preconditions: 'exam' must be from the constants file
 * Postconditions: examCollection.validityPeriods will contain the new validity period
 */
export function Validity(exam, validityString) {
  assert(Object.values(Exam).indexOf(exam) >= 0, `Unknown exam "${exam}" used in Validity`);
  examCollection.validityPeriods[exam] = validityString;
}

/**
 * Get the validity period for a given exam
 * @param {String} exam The name of the exam in question
 * Preconditions: exam must be known and must have a registered validity period
 */
export function GetExamValidity(exam) {
  assert(Object.values(Exam).indexOf(exam) >= 0, `GetExamValidity called with unknown exam: ${exam}`);
  return examCollection.validityPeriods[exam];
}
