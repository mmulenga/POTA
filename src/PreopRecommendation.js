// Singleton data, not a class
import '@/configuration';
import { GetExams, GetConditionalExams, ConditionalExam, GetExamValidity } from '@/tag';
import { Exam } from '@/constants';

/**
 * ExamSummary is a simple data struct with two properties:
 * @param exams {String[]}: a list of exams to be performed
 * @param conditionalExams {ConditionalExam[]} a list of ConditionalExams to be optionally performed
 */
export class ExamSummary {
  constructor(exams, conditionalExams) {
    this.exams = exams;
    this.conditionalExams = conditionalExams;
  }
}

/**
 * Given a list of patient conditions, return an ExamSummary object
 * The ConditionalExams returned will be prefiltered so as to not include any
 * already required exams or any redundant conditional phrases (i.e. no new exams)
 * @param {String[]} patientConditions
 * @returns {ExamSummary} Summary of examinations to be performed
 */
export function PatientExamsNeeded(patientConditions) {
  const examAggregation = new Set();
  const conditionalExamAggregation = [];

  for (let i = 0; i < patientConditions.length; i += 1) {
    const exams = GetExams(patientConditions[i]);
    for (let j = 0; j < exams.length; j += 1) {
      examAggregation.add(exams[j]);
    }
  }

  // conditionPhrasesUsed is used to dedupe conditional phrases that are equivalent
  // It may come up anytime that a single row of the grid is broken up into Tags
  const conditionPhrasesUsed = [];

  for (let i = 0; i < patientConditions.length; i += 1) {
    // Filter out any previously used condition phrases
    const conditionalExams = GetConditionalExams(patientConditions[i])
      .filter(c => !(conditionPhrasesUsed.includes(c.conditionPhrase)));
    // Add remaining phrases to conditionPhrasesUsed
    conditionPhrasesUsed.push(...conditionalExams.map(c => c.conditionPhrase));

    for (let j = 0; j < conditionalExams.length; j += 1) {
      // Make a new copy of the ConditionalExam with any preconsidered exams filtered out
      const ce = new ConditionalExam(
        conditionalExams[j].conditionPhrase,
        conditionalExams[j].exams.filter(e => !(examAggregation.has(e))),
      );
      // Only aggregate this conditionalExam if it is not entirely preconsidered
      if (ce.exams.length > 0) {
        conditionalExamAggregation.push(ce);
      }
    }
  }

  // Unconditionally perform a Group and Screen
  return new ExamSummary([Exam.GnS, ...examAggregation], conditionalExamAggregation);
}

/**
 * Get the validity period for a given exam
 * @param {String} exam The name of the exam in question
 */
export function ExamValidity(exam) {
  return GetExamValidity(exam);
}
