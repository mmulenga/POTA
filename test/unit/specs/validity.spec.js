import { Condition, Exam } from '@/constants';
import { PatientExamsNeeded, ExamValidity } from '@/PreopRecommendation';
import { GetExams, GetAllExams, GetConditionalExams, Tag, GetExamValidity, Validity } from '@/tag';

let result;

describe('Testing GetExamValidity() with all preop examinations', () => {
  it('returns the correct validity period exams for GnS', () => {
    result = GetExamValidity(Exam.GnS);
    expect(result).toEqual(expect.stringContaining('1 month'));
  });
  it('correct validity period for GnS to match snapshot', () => {
    expect(result).toMatchSnapshot();
  });

  it('returns the correct conditional exams for ECG', () => {
    result = GetExamValidity(Exam.ECG);
    expect(result).toEqual(expect.stringContaining('3 months'));
  });
  it('correct validity period for ECG to match snapshot', () => {
    expect(result).toMatchSnapshot();
  });

  it('returns the correct conditional exams for CBC', () => {
    result = GetExamValidity(Exam.CBC);
    expect(result).toEqual(expect.stringContaining('3 months'));
  });
  it('correct validity period for CBC to match snapshot', () => {
    expect(result).toMatchSnapshot();
  });

  it('returns the correct conditional exams for RenPanel', () => {
    result = GetExamValidity(Exam.RenPanel);
    expect(result).toEqual(expect.stringContaining('3 months'));
  });
  it('correct validity period for RenPanel to match snapshot', () => {
    expect(result).toMatchSnapshot();
  });

  it('returns the correct conditional exams for PTTINR', () => {
    result = GetExamValidity(Exam.PTTINR);
    expect(result).toEqual(expect.stringContaining('3 months'));
  });
  it('correct validity period for PTTINR to match snapshot', () => {
    expect(result).toMatchSnapshot();
  });

  it('returns the correct conditional exams for LFT', () => {
    result = GetExamValidity(Exam.LFT);
    expect(result).toEqual(expect.stringContaining('3 months'));
  });
  it('correct validity period for LFT to match snapshot', () => {
    expect(result).toMatchSnapshot();
  });

  it('returns the correct conditional exams for Gluc', () => {
    result = GetExamValidity(Exam.Gluc);
    expect(result).toEqual(expect.stringContaining('3 months'));
  });
  it('correct validity period for Gluc to match snapshot', () => {
    expect(result).toMatchSnapshot();
  });

  it('returns the correct conditional exams for HbA1C', () => {
    result = GetExamValidity(Exam.HbA1C);
    expect(result).toEqual(expect.stringContaining('3 months'));
  });
  it('correct validity period for HbA1C to match snapshot', () => {
    expect(result).toMatchSnapshot();
  });

  it('returns the correct conditional exams for CXR', () => {
    result = GetExamValidity(Exam.CXR);
    expect(result).toEqual(expect.stringContaining('6 months'));
  });
  it('correct validity period for CXR to match snapshot', () => {
    expect(result).toMatchSnapshot();
  });

  it('returns the correct conditional exams for TSH', () => {
    result = GetExamValidity(Exam.TSH);
    expect(result).toEqual(expect.stringContaining('6 months'));
  });
  it('correct validity period for TSH to match snapshot', () => {
    expect(result).toMatchSnapshot();
  });
});

