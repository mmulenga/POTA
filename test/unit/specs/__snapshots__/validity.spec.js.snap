// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing ExamValidity() with all preop examinations correct validity period for CBC to match snapshot 1`] = `"should be done within 3 months of surgery"`;

exports[`Testing ExamValidity() with all preop examinations correct validity period for CXR to match snapshot 1`] = `"should be done within 6 months of surgery"`;

exports[`Testing ExamValidity() with all preop examinations correct validity period for ECG to match snapshot 1`] = `"should be done within 3 months of surgery"`;

exports[`Testing ExamValidity() with all preop examinations correct validity period for Gluc to match snapshot 1`] = `"should be done within 3 months of surgery"`;

exports[`Testing ExamValidity() with all preop examinations correct validity period for GnS to match snapshot 1`] = `"should be done within 1 month of surgery (or by local lab/surgical policy)"`;

exports[`Testing ExamValidity() with all preop examinations correct validity period for HbA1C to match snapshot 1`] = `"should be done within 3 months of surgery"`;

exports[`Testing ExamValidity() with all preop examinations correct validity period for LFT to match snapshot 1`] = `"should be done within 3 months of surgery"`;

exports[`Testing ExamValidity() with all preop examinations correct validity period for PTTINR to match snapshot 1`] = `"should be done within 3 months of surgery"`;

exports[`Testing ExamValidity() with all preop examinations correct validity period for RenPanel to match snapshot 1`] = `"should be done within 3 months of surgery"`;

exports[`Testing ExamValidity() with all preop examinations correct validity period for TSH to match snapshot 1`] = `"should be done within 6 months of surgery"`;

exports[`Testing GetExamValidity() with all preop examinations correct validity period for CBC to match snapshot 1`] = `"should be done within 3 months of surgery"`;

exports[`Testing GetExamValidity() with all preop examinations correct validity period for CXR to match snapshot 1`] = `"should be done within 6 months of surgery"`;

exports[`Testing GetExamValidity() with all preop examinations correct validity period for ECG to match snapshot 1`] = `"should be done within 3 months of surgery"`;

exports[`Testing GetExamValidity() with all preop examinations correct validity period for Gluc to match snapshot 1`] = `"should be done within 3 months of surgery"`;

exports[`Testing GetExamValidity() with all preop examinations correct validity period for GnS to match snapshot 1`] = `"should be done within 1 month of surgery (or by local lab/surgical policy)"`;

exports[`Testing GetExamValidity() with all preop examinations correct validity period for HbA1C to match snapshot 1`] = `"should be done within 3 months of surgery"`;

exports[`Testing GetExamValidity() with all preop examinations correct validity period for LFT to match snapshot 1`] = `"should be done within 3 months of surgery"`;

exports[`Testing GetExamValidity() with all preop examinations correct validity period for PTTINR to match snapshot 1`] = `"should be done within 3 months of surgery"`;

exports[`Testing GetExamValidity() with all preop examinations correct validity period for RenPanel to match snapshot 1`] = `"should be done within 3 months of surgery"`;

exports[`Testing GetExamValidity() with all preop examinations correct validity period for TSH to match snapshot 1`] = `"should be done within 6 months of surgery"`;

exports[`Testing Validity() with DaleDisease correct validity period for DaleDisease to match snapshot 1`] = `"should be done within 99 months of surgery"`;

exports[`Testing Validity() with DaleDisease correct validity period for DaleDisease to match snapshot 2`] = `"should be done within 99 months of surgery"`;
